### PR TITLE
Add missed MsgPackLite unit tests

### LINKED ISSUE #127: Unit test msgpack-lite

### PR DESCRIPTION
**Pull request by nicktorwald**
Closes #127

### CODE REVIEW

File: src/main/java/org/tarantool/jdbc/SQLMsgPackLite.java [hunk 1]
```diff
@@ -14,17 +14,17 @@
     public static final SQLMsgPackLite INSTANCE = new SQLMsgPackLite();
 
     @Override
-    public void pack(Object item, OutputStream os) throws IOException {
+    public void pack(Object item, OutputStream output) throws IOException {
```
Comment (Totktonada) [inline src/main/java/org/tarantool/jdbc/SQLMsgPackLite.java]: Unrelated and unnecessary change.

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 2]
```diff
@@ -34,48 +58,71 @@
     protected static final BigInteger BI_MAX_LONG = BigInteger.valueOf(Long.MAX_VALUE);
     protected static final BigInteger BI_MAX_64BIT = BigInteger.valueOf(2).pow(64).subtract(BigInteger.ONE);
 
-    //these values are from http://wiki.msgpack.org/display/MSGPACK/Format+specification
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/MsgPackLite.java]: mentioned in the class description (IMO, it's more appropriate place).

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 3]
```diff
@@ -34,48 +58,71 @@
...
-        DataOutputStream out = new DataOutputStream(os);
+    /**
+     * Serializes an object.
+     *
+     * @param item   object to be serialized
+     * @param output target output stream
+     *
+     * @throws IOException if any I/O errors occur while writing
+     */
+    public void pack(Object item, OutputStream output) throws IOException {
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/MsgPackLite.java]: tried to reorganize commits and extract unrelated parts to separate commits

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 4]
```diff
@@ -101,108 +148,96 @@ public void pack(Object item, OutputStream os) throws IOException {
...
-                if (bb.hasArray()) {
-                    data = bb.array();
-                } else {
-                    data = new byte[bb.capacity()];
-                    bb.position();
-                    bb.limit(bb.capacity());
-                    bb.get(data);
-                }
-            }
+        } else if (item instanceof byte[]) {
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/MsgPackLite.java]: explained in https://github.com/tarantool/tarantool-java/pull/236/commits/213ca3181043083179b44095cd851af7edda4e86

I also tried to find any usages in previous versions but failed. Seems it's never been used.

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 5]
```diff
@@ -86,13 +133,13 @@ public void pack(Object item, OutputStream os) throws IOException {
         if (item == null) {
             out.write(MP_NULL);
         } else if (item instanceof Boolean) {
-            out.write(((Boolean) item).booleanValue() ? MP_TRUE : MP_FALSE);
+            out.write(((boolean) item) ? MP_TRUE : MP_FALSE);
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/MsgPackLite.java]: here, we have unnecessary type unboxing using `Boolean.booleanValue()` that is highlighted by the code inspection. And, yes, it can be more concise.

File: src/test/java/org/tarantool/MsgPackLiteDeserializeTest.java [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+        oneByteLength[0] = (byte) 0xc4;
+        oneByteLength[1] = (byte) 0xc8;
+        byte b = 0;
+        for (int i = 2; i < 202; i++) {
+            oneByteLength[i] = b++;
+        }
+        byte[] firstBinary = new byte[200];
+        System.arraycopy(oneByteLength, 2, firstBinary, 0, 200);
+        inputStream = new ByteArrayInputStream(oneByteLength);
+        assertArrayEquals(firstBinary, (byte[]) msgPackLite.unpack(inputStream));
```
Comment (Totktonada) [inline src/test/java/org/tarantool/MsgPackLiteDeserializeTest.java]: Unlike other tests we don't verify a type of return value here. However I failed to find a type that may be successfully casted to `byte[]` other then `byte[]` itself. Maybe this is okay, just requires a comment. Or let's check the type explicitly.

File: src/test/java/org/tarantool/MsgPackLiteDeserializeTest.java [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+        assertEquals(new String(fourBytesLength, 5, 80000, StandardCharsets.UTF_8), msgPackLite.unpack(inputStream));
+    }
+
+    /**
+     * Unpacks binaries.
+     *
+     * @see MsgPackLiteSerializeTest#testPackBinArrays()
+     */
+    @Test
+    void testUnpackBinArrays() throws IOException {
```
Comment (Totktonada) [inline src/test/java/org/tarantool/MsgPackLiteDeserializeTest.java]: I would add a test case that will deserialize a sequence of bytes that is not a correct UTF-8 to verify that the deserializer does not wrongly assume that data in UTF-8 and does not replace anything with the replacement character. Maybe this is tested implicitly with the large arrays below, but don't sure. Either add a comment that make it obvious or, better, add a small test case that will check it explicitly.

File: src/test/java/org/tarantool/MsgPackLiteSerializeTest.java [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+     *
+     * +--------+
+     * |0XXXXXXX|
+     * +--------+
+     *
+     * 0XXXXXXX is 8-bit unsigned integer [0..127]
+     */
+    @Test
+    void testPackFixedPositiveInteger() throws IOException {
+        msgPackLite.pack(0, outputStream);
```
Comment (Totktonada) [inline src/test/java/org/tarantool/MsgPackLiteSerializeTest.java]: We may verify byte / short / long, their boxed representations and BigInteger. It worth to add at least `byte` and `Integer` I think.

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: Thanks for the explanation. Okay so.

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: I don't see a difference between `booleanValue()` and cast to `boolean`, to be honest, however okay: it is to eliminate a warning (I mean the highlight).

Is not it would be more logical to cast to `Boolean` (boxed) type?

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: Thank you.

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: Sure.

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 9]
```diff
@@ -218,8 +288,17 @@ public void pack(Object item, OutputStream os) throws IOException {
...
+    /**
+     * De-serializes next object from the binary stream.
+     *
+     * @param inputStream source data stream
+     *
+     * @return de-serialized value
+     *
+     * @throws IOException if any I/O errors occur while reading
+     */
+    public Object unpack(InputStream inputStream) throws IOException {
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: I see now, it is noted in the class javadoc. I would only add explicitly that signed types uses smaller possible type before your note about uint8/16 (BTW, it misses positive fixint).

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 10]
```diff
@@ -236,11 +315,14 @@ public Object unpack(InputStream is) throws IOException {
         case MP_DOUBLE:
             return in.readDouble();
         case MP_UINT8:
-            return in.read(); // read single byte, return as int
+            // uint8 -> int
+            return in.readByte() & MAX_8BIT;
         case MP_UINT16:
-            return in.readShort() & MAX_16BIT; // read short, trick Java into treating it as unsigned, return int
```
Comment (Totktonada) [inline src/main/java/org/tarantool/MsgPackLite.java]: 'trick Java into treating it as unsigned' comment was useful, at least for me.

File: src/main/java/org/tarantool/MsgPackLite.java [hunk 11]
```diff
@@ -7,15 +7,59 @@
...
+ *         primitive types and their wrappers (byte, char, short, int, long, float, double, boolean) as
+ *         the corresponding MsgPack format values.
+ *     </li>
+ *     <li>{@link java.math.BigInteger}, especially to represent a MsgPack uint 64 value</li>
+ *     <li>{@link org.tarantool.Code} as an integer value</li>
+ *     <li>{@link java.lang.String}</li>
+ *     <li>java arrays and {@link java.util.List} as an MsgPack array format</li>
+ *     <li>{@link java.util.Map} as an MsgPack map format</li>
+ *     <li>
+ *         {@link java.util.concurrent.Callable} that can return any object of the types
```
Comment (Totktonada) [inline src/main/java/org/tarantool/MsgPackLite.java]: A few tests for Code and Callable would make the coverage 100% in fact (except erroneous situations). That would be good, I think.

Maybe also it would be worthful to add one invalid msgpack deserialization case.

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/MsgPackLite.java]: > Is not it would be more logical to cast to Boolean (boxed) type?

Looks redundant. 

Java ternary [operator](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.25) evaluates the first expression to `boolean` and decide either branch should be evaluated:

> ... At run time, the first operand expression of the conditional expression is evaluated first. If necessary, **unboxing** conversion is performed on the result.

> The resulting `boolean` value is then used to choose either the second or the third operand expression ...

[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java]: So, back to the change itself:

```diff
-            out.write(((Boolean) item).booleanValue() ? MP_TRUE : MP_FALSE);
+            out.write(((boolean) item) ? MP_TRUE : MP_FALSE);
```

We replaced cast to `Boolean` and unboxing with just unboxing. Okay so.

File: src/main/java/org/tarantool/MsgPackLite.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/MsgPackLite.java:251]: Now I think about this again and it seems ByteBuffer may be passed as an argument to a tarantool operation and formally this change breaks backward compatibility.

I'll unresolve it to think again.
